File: ocaml/test/test_execution.ml
open Core
open OUnit2

let test_market_data_subscription _test_ctxt =
  let engine = ExecutionEngine.create () in
  let received_data = ref None in
  let symbol = "AAPL" in

  ExecutionEngine.start engine;
  ExecutionEngine.subscribe_market_data engine symbol (fun data ->
    received_data := Some data
  );

  (* Wait a bit to receive market data *)
  Unix.sleep 1;

  match !received_data with
  | None -> assert_failure "No market data received"
  | Some data ->
    assert_equal data.symbol symbol;
    assert (data.bid > 0.0);
    assert (data.ask > data.bid);
    assert (data.volume > 0L);

  ExecutionEngine.stop engine;
  ExecutionEngine.destroy engine

let test_order_submission _test_ctxt =
  let engine = ExecutionEngine.create () in
  let symbol = "AAPL" in
  let order = {
    symbol;
    price = 150.0;
    quantity = 100;
    side = Buy;
    order_id = None;
  } in

  ExecutionEngine.start engine;
  let order_with_id = ExecutionEngine.submit_order engine order in
  assert (Option.is_some order_with_id.order_id);

  (* Wait for order processing *)
  Unix.sleep 1;

  let position = ExecutionEngine.get_position engine symbol in
  assert (position = 100);

  let avg_price = ExecutionEngine.get_average_price engine symbol in
  assert (avg_price > 0.0);

  ExecutionEngine.stop engine;
  ExecutionEngine.destroy engine

let test_trade_notification _test_ctxt =
  let engine = ExecutionEngine.create () in
  let received_trade = ref None in
  let symbol = "AAPL" in

  ExecutionEngine.start engine;
  ExecutionEngine.subscribe_trades engine symbol (fun trade ->
    received_trade := Some trade
  );

  (* Submit matching orders *)
  let buy_order = {
    symbol;
    price = 150.0;
    quantity = 100;
    side = Buy;
    order_id = None;
  } in
  let sell_order = {
    symbol;
    price = 150.0;
    quantity = 100;
    side = Sell;
    order_id = None;
  } in

  ignore (ExecutionEngine.submit_order engine buy_order);
  ignore (ExecutionEngine.submit_order engine sell_order);

  (* Wait for trade processing *)
  Unix.sleep 1;

  match !received_trade with
  | None -> assert_failure "No trade received"
  | Some trade ->
    assert_equal trade.symbol symbol;
    assert_equal trade.quantity 100;
    assert (trade.price >= 150.0);

  ExecutionEngine.stop engine;
  ExecutionEngine.destroy engine

let test_pnl_calculation _test_ctxt =
  let engine = ExecutionEngine.create () in
  let symbol = "AAPL" in

  ExecutionEngine.start engine;

  (* Submit a buy order *)
  let buy_order = {
    symbol;
    price = 150.0;
    quantity = 100;
    side = Buy;
    order_id = None;
  } in
  ignore (ExecutionEngine.submit_order engine buy_order);

  (* Wait for order processing *)
  Unix.sleep 1;

  let unrealized_pnl = ExecutionEngine.get_unrealized_pnl engine symbol in
  let realized_pnl = ExecutionEngine.get_realized_pnl engine symbol in

  assert (unrealized_pnl <> 0.0);  (* Should have some P&L due to market moves *)
  assert_equal realized_pnl 0.0;   (* No realized P&L yet *)

  (* Submit a sell order to realize P&L *)
  let sell_order = {
    symbol;
    price = 151.0;
    quantity = 100;
    side = Sell;
    order_id = None;
  } in
  ignore (ExecutionEngine.submit_order engine sell_order);

  (* Wait for order processing *)
  Unix.sleep 1;

  let final_realized_pnl = ExecutionEngine.get_realized_pnl engine symbol in
  assert (final_realized_pnl > 0.0);  (* Should have realized profit *)

  ExecutionEngine.stop engine;
  ExecutionEngine.destroy engine

let suite =
  "execution_engine_test" >::: [
    "test_market_data_subscription" >:: test_market_data_subscription;
    "test_order_submission" >:: test_order_submission;
    "test_trade_notification" >:: test_trade_notification;
    "test_pnl_calculation" >:: test_pnl_calculation;
  ]

let () =
  run_test_tt_main suite

File: cpp/test/test_execution.cpp
#include "execution_engine.hpp"
#include <iostream>
#include <iomanip>
#include <thread>
#include <chrono>
#include <vector>
#include <sstream>
#include <random>

// ANSI color codes for prettier output
namespace Color {
    const char* Reset   = "\033[0m";
    const char* Red     = "\033[31m";
    const char* Green   = "\033[32m";
    const char* Yellow  = "\033[33m";
    const char* Blue    = "\033[34m";
    const char* Magenta = "\033[35m";
}

void display_position(const std::string& symbol, const trading::ExecutionEngine& engine) {
    std::cout << Color::Yellow << "[Position] " << Color::Reset
              << symbol << ": " << engine.get_position(symbol)
              << " @ $" << std::fixed << std::setprecision(2) << engine.get_average_price(symbol)
              << " P&L: $" << engine.get_realized_pnl(symbol) 
              << " (Unrealized: $" << engine.get_unrealized_pnl(symbol) << ")"
              << std::endl;
}

int main() {
    std::cout << "Starting Execution Engine Test...\n" << std::endl;

    trading::ExecutionEngine engine;
    std::cout << "Created engine\n";
    
    engine.start();
    std::cout << "Started engine\n";

    const std::string symbol = "AAPL";
    std::cout << "Testing with symbol: " << symbol << std::endl;

    // Subscribe to market data and trades
    engine.subscribe_market_data(symbol, 
        [](const trading::MarketData& data) {
            std::cout << Color::Blue << "[Market] " << Color::Reset
                      << data.symbol << " @ $" << std::fixed << std::setprecision(2) 
                      << data.price << std::endl;
        }
    );

    engine.subscribe_trades(symbol,
        [](const trading::Trade& trade) {
            std::cout << Color::Green << "[Trade] " << Color::Reset
                      << trade.symbol << " @ $" << std::fixed << std::setprecision(2) 
                      << trade.price << " x " << trade.quantity << std::endl;
        }
    );

    std::cout << "\nStarting Position:\n";
    display_position(symbol, engine);

    // First trade: Buy 100 shares at $100
    std::cout << "\nSubmitting buy order (100 shares @ $100.00)...\n";
    trading::Order buy_order{
        "", 
        symbol,
        100.0,
        100,
        true
    };
    engine.submit_order(buy_order);
    
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
    std::cout << "\nPosition after buy:\n";
    display_position(symbol, engine);

    // Second trade: Sell 50 shares at $105
    std::cout << "\nSubmitting sell order (50 shares @ $105.00)...\n";
    trading::Order sell_order1{
        "",
        symbol,
        105.0,
        50,
        false
    };
    engine.submit_order(sell_order1);
    
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
    std::cout << "\nPosition after partial sell:\n";
    display_position(symbol, engine);

    // Third trade: Sell remaining 50 shares at $110
    std::cout << "\nSubmitting sell order (50 shares @ $110.00)...\n";
    trading::Order sell_order2{
        "",
        symbol,
        110.0,
        50,
        false
    };
    engine.submit_order(sell_order2);
    
    std::this_thread::sleep_for(std::chrono::milliseconds(100));
    std::cout << "\nFinal Position:\n";
    display_position(symbol, engine);

    std::cout << "\nStopping engine...\n";
    engine.stop();
    std::cout << "Test completed.\n" << std::endl;

    return 0;
}
